### PR TITLE
Tweak cache pidlocking

### LINKED ISSUE #50206: Cache generation for stdlib on read-only FS fails
```
Failed to precompile OrdinaryDiffEq [1dea7af3-3e70-54e6-95c3-0bf5283fa5ed] to "/home/pkgeval/.julia/compiled/v1.10/OrdinaryDiffEq/jl_uICwpR".
ERROR: LoadError: IOError: open("/opt/julia/share/julia/stdlib/v1.10/DelimitedFiles/src/DelimitedFiles.jl.pidlock", 194, 292): read-only file system (EROFS)
Stacktrace:
  [1] uv_error(prefix::String, c::Int32)
    @ Base ./libuv.jl:100 [inlined]
  [2] open(path::String, flags::UInt16, mode::UInt16)
    @ Base.Filesystem ./filesystem.jl:121
  [3] tryopen_exclusive(path::String, mode::UInt16)
    @ FileWatching.Pidfile /opt/julia/share/julia/stdlib/v1.10/FileWatching/src/pidfile.jl:210
  [4] open_exclusive(path::String; mode::UInt16, poll_interval::Int64, wait::Bool, stale_age::Int64)
    @ FileWatching.Pidfile /opt/julia/share/julia/stdlib/v1.10/FileWatching/src/pidfile.jl:236
  [5] mkpidlock(at::String, pid::Int32; stale_age::Int64, refresh::Float64, kwopts::@Kwargs{wait::Bool})
    @ FileWatching.Pidfile /opt/julia/share/julia/stdlib/v1.10/FileWatching/src/pidfile.jl:66
  [6] mkpidlock(f::Base.var"#962#963"{Base.PkgId}, at::String, pid::Int32; kwopts::@Kwargs{wait::Bool})
    @ FileWatching.Pidfile /opt/julia/share/julia/stdlib/v1.10/FileWatching/src/pidfile.jl:90
  [7] mkpidlock(f::Function, at::String; kwopts::@Kwargs{wait::Bool})
    @ FileWatching.Pidfile /opt/julia/share/julia/stdlib/v1.10/FileWatching/src/pidfile.jl:87 [inlined]
  [8] trymkpidlock(::Function, ::Vararg{Any}; kwargs::@Kwargs{})
    @ FileWatching.Pidfile /opt/julia/share/julia/stdlib/v1.10/FileWatching/src/pidfile.jl:110
  [9] trymkpidlock(::Function, ::Vararg{Any})
    @ FileWatching.Pidfile /opt/julia/share/julia/stdlib/v1.10/FileWatching/src/pidfile.jl:108
 [10] invokelatest(::Any, ::Returns{Nothing}, ::String; kwargs::@Kwargs{})
    @ Base ./essentials.jl:864 [inlined]
 [11] invokelatest(::Any, ::Base.var"#962#963"{Base.PkgId}, ::String)
    @ Base ./essentials.jl:861 [inlined]
 [12] maybe_cachefile_lock(f::Base.var"#962#963"{Base.PkgId}, pkg::Base.PkgId, srcpath::String)
    @ Base ./loading.jl:2826
 [13] _require(pkg::Base.PkgId, env::String)
    @ Base ./loading.jl:1905
 [14] __require_prelocked(uuidkey::Base.PkgId, env::String)
    @ Base ./loading.jl:1747
 [15] invoke_in_world(::UInt64, 
```

Seen during https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/d9a5a29_vs_65523e4/MultiStateSystems.primary.log

cc: @IanButterworth

### LINKED ISSUE #50213: Check if pidlock process is still running
I somehow interrupted a precompilation process, resulting in a stale pidlock somewhere in my depot. This causes precompilation to hang, which for all intents and purposes looks like a plain hang (IIUC https://github.com/JuliaLang/Pkg.jl/pull/3519 wouldn't help here, because this is triggered by `using Package` in code, not by `Pkg.precompile()`).

Setting JULIA_DEBUG shows that its stuck waiting for a pidlock, as does the backtrace when interrupting:

```
┌ Debug: Waiting for another process (pid: 5719) to finish precompiling CUDA [052768ef-5323-5732-b1bb-66c8b64840ba]
└ @ Base loading.jl:2831
^C
[7169] signal (2): Interrupt
in expression starting at /home/tim/Julia/pkg/CUDA/wip.jl:1
epoll_wait at /usr/lib/libc.so.6 (unknown line)
uv__io_poll at /workspace/srcdir/libuv/src/unix/epoll.c:236
uv_run at /workspace/srcdir/libuv/src/unix/core.c:400
ijl_task_get_next at /home/tim/Julia/src/julia/src/partr.c:436
unknown function (ip: 0x7f59f774c93e)
unknown function (ip: 0x7f59f75d646c)
unknown function (ip: 0x7f59f77a27c5)
unknown function (ip: 0x7f59f704af20)
unknown function (ip: 0x7f59f6fa0c03)
watch_file at /home/tim/Julia/src/julia/build/dev/usr/share/julia/stdlib/v1.10/FileWatching/src/FileWatching.jl:778 [inlined]
JuliaLang/Pkg.jl#13 at /home/tim/Julia/src/julia/build/dev/usr/share/julia/stdlib/v1.10/FileWatching/src/pidfile.jl:257
unknown function (ip: 0x7f5a0d07daf9)
_jl_invoke at /home/tim/Julia/src/julia/src/gf.c:2870 [inlined]
ijl_apply_generic at /home/tim/Julia/src/julia/src/gf.c:3071
jl_apply at /home/tim/Julia/src/julia/src/julia.h:1963 [inlined]
start_task at /home/tim/Julia/src/julia/src/task.c:1238
unknown function (ip: (nil))
Allocations: 2836 (Pool: 2827; Big: 9); GC: 0
```

Crucially, process 5719 just doesn't exist anymore. Instead of waiting for the pidlock to go stale, which seems like ages (it's been almost 10 minutes), shouldn't we first check if the process even exists and if not break the lock?

### PR DESCRIPTION
**Pull request by IanButterworth**
Fixes https://github.com/JuliaLang/julia/issues/50206

Moves the pidfile from the package source to the depot. The pidlock is now specific to the cachefile but ignoring any preferences (which would change the cachefile path) as the prefs hash can only be determined after precompilation.

@vchuravy 

Fixes https://github.com/JuliaLang/julia/issues/50213

Stale pidfile rejection was unintentionally disabled, `stale_age` should have been set. I'm proposing 60 seconds. See comments for explanation.

@maleadt

### CODE REVIEW

File: base/loading.jl [hunk 1]
```diff
@@ -2819,11 +2819,20 @@ global mkpidlock_hook
 global trymkpidlock_hook
 global parse_pidfile_hook
 
-# allows processes to wait if another process is precompiling a given source already
-function maybe_cachefile_lock(f, pkg::PkgId, srcpath::String)
+# The perferences hash is only known after precompilation so just assume no preferences
```
Comment (barucden) [inline base/loading.jl:2822]:
```suggestion
# The preferences hash is only known after precompilation so just assume no preferences
```

File: base/loading.jl [hunk 2]
```diff
@@ -2819,11 +2819,20 @@ global mkpidlock_hook
...
+# meaning that if all other conditions are equal, the same package cannot be precompiled
+# with different preferences at the same time.
+compilecache_pidfile_path(pkg::PkgId) = compilecache_path(pkg, UInt64(0)) * ".pidfile"
+
+# Allows processes to wait if another process is precompiling a given source already.
+# The lock file is deleted and precompilation will proceed after `stale_age` seconds if
+#  - the locking process no longer exists
+#  - the lock is held by another host, since processes cannot be checked remotely
+# or after `stale_age * 25` seconds if it does still exist.
+function maybe_cachefile_lock(f, pkg::PkgId, srcpath::String; stale_age=60)
```
Comment (vtjnash) [inline base/loading.jl:2832]: This is probably too short. It is supposed to be a lower bound, not an upper bound. Something around 300 is probably okay (5 minutes) though, or perhaps even 600.

File: base/loading.jl [hunk 3]
```diff
@@ -2819,11 +2819,20 @@ global mkpidlock_hook
 global trymkpidlock_hook
 global parse_pidfile_hook
 
-# allows processes to wait if another process is precompiling a given source already
-function maybe_cachefile_lock(f, pkg::PkgId, srcpath::String)
+# The preferences hash is only known after precompilation so just assume no preferences
+# meaning that if all other conditions are equal, the same package cannot be precompiled
+# with different preferences at the same time.
+compilecache_pidfile_path(pkg::PkgId) = compilecache_path(pkg, UInt64(0)) * ".pidfile"
```
Comment (vtjnash) [inline base/loading.jl:2825]: This also uses the path to `Base.active_project`, which is also undesirable for the same reason as preferences. Can you make that an argument there, and pass `""` here to ignore it for this lock?

**Comment by vtjnash** [comment]
Thanks for fixing those. Sorry for the slow review!

File: base/loading.jl (hunk 2 above)
Comment (IanButterworth) [reply base/loading.jl:2832]: Changed in https://github.com/JuliaLang/julia/pull/50254

File: base/loading.jl (hunk 3 above)
Comment (IanButterworth) [reply base/loading.jl:2825]: Changed in https://github.com/JuliaLang/julia/pull/50254

**Pull request merged**